From cfed1b0018b36ae2c1c39a6933b4285779b183b4 Mon Sep 17 00:00:00 2001
From: Shilin Victor <chrono.monochrome@gmail.com>
Date: Mon, 20 Aug 2018 22:57:43 +0300
Subject: [PATCH 269/296] schedule vsync immediately when requested from the
 Looper thread

This patch allows Choreographer to schedule vsync immediately when it
is requested from the Looper thread. Previously this is not allowed
when vsync is already scheduled from a binder thread and hence there
is a delay up to 16ms (one vsync interval) in processing frame
callbacks.
---
 core/java/android/view/Choreographer.java | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/core/java/android/view/Choreographer.java b/core/java/android/view/Choreographer.java
index 597be686449..b1ab036a6a5 100644
--- a/core/java/android/view/Choreographer.java
+++ b/core/java/android/view/Choreographer.java
@@ -594,6 +594,19 @@ public final class Choreographer {
                 msg.setAsynchronous(true);
                 mHandler.sendMessageAtTime(msg, nextFrameTime);
             }
+        } else {
+            if (USE_VSYNC) {
+                if (mHandler.hasMessages(MSG_DO_SCHEDULE_VSYNC)) {
+                    // If running on the Looper thread, then schedule the vsync immediately.
+                    if (isRunningOnLooperThreadLocked()) {
+                        if (DEBUG_FRAMES) {
+                            Log.d(TAG, "Scheduling next frame on vsync.");
+                        }
+                        scheduleVsyncLocked();
+                        mHandler.removeMessages(MSG_DO_SCHEDULE_VSYNC);
+                    }
+                }
+            }
         }
     }
 
-- 
2.11.0

